pallets/capacity/src/lib.rs

```diff
@@ -518,19 +517,81 @@ impl<T: Config> Pallet<T> {
 		unstaker: &T::AccountId,
 		amount: BalanceOf<T>,
 	) -> Result<BalanceOf<T>, DispatchError> {
-		let mut staking_account =
-			Self::get_staking_account_for(unstaker).ok_or(Error::<T>::NotAStakingAccount)?;
-		ensure!(amount <= staking_account.active, Error::<T>::AmountToUnstakeExceedsAmountStaked);
+		StakingAccountLedger::<T>::try_mutate_exists(
+			unstaker,
+			|maybe_staking_account| -> Result<BalanceOf<T>, DispatchError> {
+				let mut staking_account =
+					maybe_staking_account.take().ok_or(Error::<T>::NotAStakingAccount)?;
+				ensure!(
+					amount <= staking_account.active,
+					Error::<T>::AmountToUnstakeExceedsAmountStaked
+				);
+
+				let actual_unstaked_amount = staking_account.withdraw(amount)?;
+				if staking_account.active.is_zero() {
+					*maybe_staking_account = None;
+				} else {
+					*maybe_staking_account = Some(staking_account);
+				}
+				Ok(actual_unstaked_amount)
+			},
+		)
+	}
 
+	fn add_unlock_chunk(
+		unstaker: &T::AccountId,
+		actual_unstaked_amount: BalanceOf<T>,
+	) -> Result<(), DispatchError> {
 		let current_epoch: T::EpochNumber = Self::get_current_epoch();
 		let thaw_at =
 			current_epoch.saturating_add(T::EpochNumber::from(T::UnstakingThawPeriod::get()));
+		let mut unlocks = Self::get_unstake_unlocking_for(unstaker).unwrap_or_default();
+		unlocks.add(actual_unstaked_amount, thaw_at)?;
+		UnstakeUnlocks::<T>::set(unstaker, Some(unlocks));
+		Ok(())
+	}
 
-		let unstake_result = staking_account.withdraw(amount, thaw_at)?;
-
-		Self::set_staking_account(&unstaker, &staking_account);
+	// Calculates a stakable amount from a proposed amount.
+	pub(crate) fn get_stakable_amount_for(
+		staker: &T::AccountId,
+		proposed_amount: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		let account_balance = T::Currency::free_balance(&staker);
+		account_balance
+			.saturating_sub(T::MinimumTokenBalance::get())
+			.min(proposed_amount)
+	}
 
-		Ok(unstake_result)
+	pub(crate) fn do_withdraw_unstaked(
+		staker: &T::AccountId,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		let current_epoch = Self::get_current_epoch();
+		let mut amount_withdrawn: BalanceOf<T> = Zero::zero();
+		let mut total_unlocking: BalanceOf<T> = Zero::zero();
+		UnstakeUnlocks::<T>::try_mutate_exists(
```

Comment:
The logic does change; if they unstaked everything, then their staking account will have been reaped on `unstake` call.  So by this time it would not exist because unlocks are now separated.  Therefore we do not want to error if there is no staking account.